File: realtime/ios/webrtc/Sources/Realtime/Events/ServerEvent.swift
import Foundation

enum ServerEvent: Equatable {
    case outputTextDelta(String)
    case transcriptDone(String?)
    case inputTranscriptionDelta(String)
    case inputTranscriptionCompleted(String)
    case speechStarted
    case outputItemAdded
    case responseDone
    case backchannelAudioDelta(base64PCM16: String)
    case backchannelAudioDone(phrase: String?)
    case backchannelSkipped(reason: String)
    case error(String)
    case unknown(type: String)

    static func decode(_ data: Data) -> ServerEvent {
        guard let obj = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
              let type = obj["type"] as? String else {
            return .unknown(type: "")
        }
        switch type {
        case "response.output_text.delta",
             "response.text.delta",
             "response.audio_transcript.delta",
             "response.output_audio_transcript.delta":
            return .outputTextDelta(obj["delta"] as? String ?? "")
        case "response.output_audio_transcript.done":
            return .transcriptDone(obj["transcript"] as? String)
        case "response.content_part.done":
            let part = obj["part"] as? [String: Any]
            return .transcriptDone(part?["transcript"] as? String)
        case "conversation.item.input_audio_transcription.delta":
            return .inputTranscriptionDelta(obj["delta"] as? String ?? "")
        case "conversation.item.input_audio_transcription.completed":
            return .inputTranscriptionCompleted(obj["transcript"] as? String ?? "")
        case "input_audio_buffer.speech_started":
            return .speechStarted
        case "response.output_item.added":
            return .outputItemAdded
        case "response.done":
            return .responseDone
        case "response.backchannel.audio.delta":
            return .backchannelAudioDelta(base64PCM16: obj["delta"] as? String ?? "")
        case "response.backchannel.audio.done":
            return .backchannelAudioDone(phrase: obj["phrase"] as? String)
        case "response.backchannel.skipped":
            return .backchannelSkipped(reason: obj["reason"] as? String ?? "")
        case "error":
            let error = obj["error"] as? [String: Any]
            let message = error?["message"] as? String ?? obj["message"] as? String ?? "unknown error"
            return .error(message)
        default:
            return .unknown(type: type)
        }
    }
}

File: realtime/android/webrtc/app/src/main/java/ai/inworld/voiceagent/realtime/events/ServerEvent.kt
package ai.inworld.voiceagent.realtime.events

import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

sealed interface ServerEvent {
    data class OutputTextDelta(val delta: String) : ServerEvent
    data class TranscriptDone(val transcript: String?) : ServerEvent
    data class InputTranscriptionDelta(val delta: String) : ServerEvent
    data class InputTranscriptionCompleted(val transcript: String) : ServerEvent
    data object SpeechStarted : ServerEvent
    data object OutputItemAdded : ServerEvent
    data object ResponseDone : ServerEvent
    data class BackchannelAudioDelta(val base64Pcm16: String) : ServerEvent
    data class BackchannelAudioDone(val phrase: String?) : ServerEvent
    data class BackchannelSkipped(val reason: String) : ServerEvent
    data class Error(val message: String) : ServerEvent
    data class Unknown(val type: String) : ServerEvent

    companion object {
        fun decode(text: String): ServerEvent {
            val obj = runCatching { EventJson.parseToJsonElement(text).jsonObject }.getOrNull()
                ?: return Unknown("")
            val type = obj["type"]?.jsonPrimitive?.contentOrNull ?: return Unknown("")

            fun str(key: String): String? =
                runCatching { obj[key]?.jsonPrimitive?.contentOrNull }.getOrNull()

            return when (type) {
                "response.output_text.delta",
                "response.text.delta",
                "response.audio_transcript.delta",
                "response.output_audio_transcript.delta",
                -> OutputTextDelta(str("delta") ?: "")

                "response.output_audio_transcript.done" -> TranscriptDone(str("transcript"))

                "response.content_part.done" -> {
                    val part = runCatching { obj["part"]?.jsonObject }.getOrNull()
                    TranscriptDone(runCatching { part?.get("transcript")?.jsonPrimitive?.contentOrNull }.getOrNull())
                }

                "conversation.item.input_audio_transcription.delta" ->
                    InputTranscriptionDelta(str("delta") ?: "")

                "conversation.item.input_audio_transcription.completed" ->
                    InputTranscriptionCompleted(str("transcript") ?: "")

                "input_audio_buffer.speech_started" -> SpeechStarted
                "response.output_item.added" -> OutputItemAdded
                "response.done" -> ResponseDone

                "response.backchannel.audio.delta" -> BackchannelAudioDelta(str("delta") ?: "")
                "response.backchannel.audio.done" -> BackchannelAudioDone(str("phrase"))
                "response.backchannel.skipped" -> BackchannelSkipped(str("reason") ?: "")

                "error" -> {
                    val error = runCatching { obj["error"]?.jsonObject }.getOrNull()
                    val message = runCatching { error?.get("message")?.jsonPrimitive?.contentOrNull }.getOrNull()
                        ?: str("message")
                        ?: "unknown error"
                    Error(message)
                }

                else -> Unknown(type)
            }
        }
    }
}
